webpack/assets/javascripts/react_app/components/ColumnSelector/ColumnSelector.js

```diff
@@ -0,0 +1,200 @@
+          return column.key;
+        }
+        return null;
+      })
+      .filter(item => item);
+    return keys;
+  };
+
+  async function updateTablePreference() {
+    if (!tablePreference) {
+      await API.post(url, { name: 'hosts', columns: getColumnKeys() });
+      setTablePreference(true);
+    } else {
+      await API.put(`${url}/${controller}`, { columns: getColumnKeys() });
+    }
+  }
+
+  const filterItems = (item, checkedItem) => {
+    if (item.key === checkedItem.key) {
+      return true;
+    }
+
+    if (item.children) {
+      item.children = item.children
+        .map(opt => Object.assign({}, opt))
+        .filter(column => filterItems(column, checkedItem));
+      return item.children;
+    }
+
+    return null;
+  };
+
+  const flattenTree = tree => {
+    let result = [];
+    tree.forEach(item => {
+      result.push(item);
+      if (item.children) {
+        result = result.concat(flattenTree(item.children));
+      }
+    });
+    return result;
+  };
+
+  const toggleModal = () => {
+    setSelectedColumns(initialColumns);
+    setModalOpen(!isModalOpen);
+  };
+
+  const isDisabled = () => {
+    const hasPartialCheck = selectedColumns.map(
+      item => item.checkProps.checked
+    );
+    return hasPartialCheck.every(el => el === false);
+  };
+
+  const updateCheckBox = (treeViewItem, checked = true) => {
+    treeViewItem.checkProps.checked = checked;
+    if (treeViewItem.children) {
+      treeViewItem.children.forEach(item => {
+        item.checkProps.checked = checked;
+      });
+    }
+  };
+
+  const onCheck = (evt, treeViewItem) => {
+    const { checked } = evt.target;
+    const checkedItemTree = selectedColumns
+      .map(column => Object.assign({}, column))
+      .filter(item => filterItems(item, treeViewItem));
+    const flatCheckedItems = flattenTree(checkedItemTree);
+
+    if (checked) {
+      updateCheckBox(treeViewItem);
+      setSelectedColumns(
+        selectedColumns
+          .concat(
+            flatCheckedItems.filter(
+              item => !selectedColumns.some(i => i.key === item.key)
+            )
+          )
+          .filter(item => item.children)
+      );
+    } else {
+      updateCheckBox(treeViewItem, false);
+      setSelectedColumns(
+        selectedColumns.filter(item =>
+          flatCheckedItems.some(i => i.key === item.key)
+        )
+      );
+    }
+    selectedColumns.map(category => areDescendantsChecked(category));
+  };
+
+  const isChecked = dataItem => dataItem.checkProps.checked;
+  const areDescendantsChecked = dataItem => {
+    if (dataItem.children) {
+      if (dataItem.children.every(child => isChecked(child))) {
+        dataItem.checkProps.checked = true;
+      } else if (dataItem.children.some(child => isChecked(child))) {
+        dataItem.checkProps.checked = null;
+      } else {
+        dataItem.checkProps.checked = false;
+      }
+    }
+  };
+
+  if (!tablePreference) {
+    updateTablePreference();
+  }
```

Comment:
We don't need this at all, `updateTablePreference` has the needed logic already.
```suggestion
```